src/adapter/rust_type_name.rs

```diff
@@ -0,0 +1,464 @@
+        // according to docs, `len` is not guaranteed to roundtrip
+        rustdoc_types::Type::Array { type_, len } => write!(f, "[{}; {}]", Type(type_), len),
+        rustdoc_types::Type::ImplTrait(vec) => {
+            write!(f, "impl ")?;
+            intersperse_with(f, " + ", vec, GenericBound)
+        }
+        rustdoc_types::Type::Infer => write!(f, "_"),
+        rustdoc_types::Type::RawPointer { is_mutable, type_ } => {
+            let kind = if *is_mutable { "mut" } else { "const" };
+
+            write!(f, "*{kind} {}", Type(type_))
+        }
+        rustdoc_types::Type::BorrowedRef {
+            lifetime,
+            is_mutable,
+            type_,
+        } => {
+            write!(f, "&")?;
+            if let Some(lt) = lifetime {
+                write!(f, "{lt} ")?;
+            }
+
+            if *is_mutable {
+                write!(f, "mut ")?;
+            }
+
+            write!(f, "{}", Type(type_))
+        }
+        rustdoc_types::Type::QualifiedPath {
+            name,
+            args,
+            self_type,
+            trait_,
+        } => {
+            if let Some(trait_) = trait_ {
+                write!(f, "<{} as {}>", Type(self_type), Path(trait_))?;
+            } else {
+                write!(f, "{}", Type(self_type))?;
+            }
+
+            write!(f, "::{}{}", name, GenericArgs(args))
+        }
+        rustdoc_types::Type::Pat { .. } => unimplemented!("Type::Pat is unstable"),
+    }
+}
+
+display_wrapper!(Type, fmt_type);
+
+fn fmt_generic_bound(this: &GenericBound, f: &mut Formatter<'_>) -> Result {
+    match &this.0 {
+        rustdoc_types::GenericBound::TraitBound {
+            trait_,
+            generic_params,
+            modifier,
+        } => {
+            match modifier {
+                rustdoc_types::TraitBoundModifier::Maybe => write!(f, "?")?,
+                // TODO: check this/find a good reference.  it's currently unstablke
+                rustdoc_types::TraitBoundModifier::MaybeConst => write!(f, "~const ")?,
+                rustdoc_types::TraitBoundModifier::None => (),
+            };
+
+            write!(f, "{}", Path(trait_))?;
+
+            if generic_params.is_empty() {
+                write!(f, "<")?;
+                intersperse_with(f, ", ", generic_params, GenericParamDef)?;
+                write!(f, ">")?;
+            }
+
+            Ok(())
+        }
+        rustdoc_types::GenericBound::Outlives(lt) => write!(f, "{lt}"),
+        rustdoc_types::GenericBound::Use(vec) => {
+            write!(f, "use<")?;
+            intersperse(f, ", ", vec, String::fmt)?;
+            write!(f, ">")
+        }
+    }
+}
+
+display_wrapper!(GenericBound, fmt_generic_bound);
+
+fn fmt_constant(this: &Constant, f: &mut Formatter<'_>) -> Result {
+    if let Some(val) = &this.0.value {
+        write!(f, "{}", val)
+    } else {
+        // TODO: investigate expr docs - stringified is unstable
```

Comment:
Consider elaborating on this TODO note. Ideally, we'd like to make it clear what the action item is and why it's important, in case we decide it isn't worth addressing immediately and then someone else down the line inherits this code.